src/community/security/oauth2/oauth2-core/src/main/java/org/geoserver/security/oauth2/GeoServerOAuthAuthenticationFilter.java

```diff
@@ -145,22 +147,30 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
                     || authentication == null
                     || authentication instanceof AnonymousAuthenticationToken) {
 
+                // performs full authentication (token validation, roles calculation, etc.)
                 doAuthenticate((HttpServletRequest) request, (HttpServletResponse) response);
 
                 Authentication postAuthentication =
                         SecurityContextHolder.getContext().getAuthentication();
                 if (postAuthentication != null) {
-                    if (cacheAuthentication(postAuthentication, (HttpServletRequest) request)) {
-                        getSecurityManager()
-                                .getAuthenticationCache()
-                                .put(getName(), getCacheKey((HttpServletRequest) request), postAuthentication);
-                    }
+                    tryCacheAuthentication((HttpServletRequest) request, cacheKey, postAuthentication);
```

Comment:
Heads up... the `authenticateFromCache` form the `GeoServerSecurityFilter` class seems to be returning a `NULL` `cacheKey` whenever the `getSecurityManager().getAuthenticationCache().get(getName(), cacheKey)` is `HIT`... could this have an impact on the code? The `tryCacheAuthentication` could throw an exception if the `cacheKey` is null?

I don't fully understand the `authenticateFromCache` logic, but we should problably also check for the `cacheKey` validity here.

Reply:
It's actually going to work out. If the cache is hit we don't need to authenticate again, and the authentication is set inside the SecurityContextHolder, so the cacheKey is not used any more. 
That said, just for extra safety, I've guarded the "tryCache" with a non-null check and clarified with a comment what is going on.